src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -291,6 +401,212 @@ export async function putDataInTextureThenDrawAndCheckResults<T extends Dimensio
+  });
+  t.trackForCleanup(renderTarget);
+
+  const size = reifyExtent3D(texture.descriptor.size);
+  const ddx = ((options.ddx ?? 1) * renderTarget.width) / size.width;
+  const ddy = ((options.ddy ?? 1) * renderTarget.height) / size.height;
+  const ddxStart = ((options.ddxStart ?? 0) * renderTarget.width) / size.width;
+  const ddyStart = ((options.ddyStart ?? 0) * renderTarget.height) / size.height;
+  const offset = options.offset;
+
+  const code = `
+struct InOut {
+  @builtin(position) pos: vec4f,
+  @location(0) uv: vec2f,
+};
+
+@vertex fn vs(@builtin(vertex_index) vertex_index : u32) -> InOut {
+  let positions = array(
+    vec2f(-1,  1), vec2f( 1,  1),
+    vec2f(-1, -1), vec2f( 1, -1),
+  );
+  let pos = positions[vertex_index];
+  return InOut(
+    vec4f(pos, 0, 1),
+    (pos * 0.5 + 0.5) * vec2f(${ddx}, ${ddy}) - vec2f(${ddxStart}, ${ddyStart}),
+  );
+}
+
+@group(0) @binding(0) var          T    : texture_2d<f32>;
+@group(0) @binding(1) var          S    : sampler;
+
+@fragment fn fs(v: InOut) -> @location(0) vec4f {
+  return textureSample(T, S, v.uv${offset ? `, vec2i(${offset[0]},${offset[1]})` : ''});
+}
+`;
+
+  const shaderModule = device.createShaderModule({ code });
+
+  const pipeline = device.createRenderPipeline({
+    layout: 'auto',
+    vertex: { module: shaderModule },
+    fragment: {
+      module: shaderModule,
+      targets: [{ format }],
+    },
+    primitive: { topology: 'triangle-strip' },
+  });
+
+  const gpuTexture = createTextureFromTexelViews(device, texture.texels, texture.descriptor);
+  t.trackForCleanup(gpuTexture);
+  const gpuSampler = device.createSampler(sampler);
+
+  const bindGroup = device.createBindGroup({
+    layout: pipeline.getBindGroupLayout(0),
+    entries: [
+      { binding: 0, resource: gpuTexture.createView() },
+      { binding: 1, resource: gpuSampler },
+    ],
+  });
+
+  const encoder = device.createCommandEncoder();
+
+  const renderPass = encoder.beginRenderPass({
+    colorAttachments: [{ view: renderTarget.createView(), loadOp: 'clear', storeOp: 'store' }],
+  });
+
+  renderPass.setPipeline(pipeline);
+  renderPass.setBindGroup(0, bindGroup);
+  renderPass.draw(4);
+  renderPass.end();
+  device.queue.submit([encoder.finish()]);
+
+  const expTexelView = softwareRasterize<T>(
+    texture,
+    sampler,
+    [renderTarget.width, renderTarget.height],
+    { ddx, ddy, ddxStart, ddyStart, offset }
+  );
+
+  // Note: I'm not sure what we should do here. My assumption is, given texels
+  // have random values, the difference between 2 texels can be very large. In
+  // the current version, for a float texture they can be +/- 1000 difference.
+  // Sampling is very GPU dependent. So if one pixel gets a random value of
+  // -1000 and the neighboring pixel gets +1000 then any slight variation in how
+  // sampling is applied will generate a large difference when interpolating
+  // between -1000 and +1000.
```

Comment:
Based on reading this comment I think what you have is fine. Would it actually help if the random numbers weren't as widely spread? I think it would be fine to constrain them more but I'm not sure it actually reduces the tolerances you need.

Reply:
The one time I tried this I had to lower them all the way to +2/-2 for whatever tolerances I originally had. I don't know if +2/-2 is better (Because we use a lower tolerance) or just more of the same.  The dEQP is quantizing to 8 bits so I think that makes up much of the difference.